expr/functions.go

```diff
@@ -58,407 +70,486 @@ var allFns = map[string]struct {
+		f, _ := zng.DecodeFloat64(zv.Bytes)
+		return zng.Value{zv.Type, args.Float64(math.Round(f))}, nil
 
-	case zng.IdFloat64:
-		return zngnative.Value{zng.TypeFloat64, math.Round(args[0].Value.(float64))}, nil
-
-	default:
+	}
+	if !zng.IsNumber(id) {
 		return err("Math.round", ErrBadArgument)
 	}
+	return zv, nil
 }
 
-func mathPow(args []zngnative.Value) (zngnative.Value, error) {
-	x, ok := zngnative.CoerceNativeToFloat64(args[0])
+func mathPow(args *Args) (zng.Value, error) {
+	x, ok := CoerceToFloat(args.vals[0])
 	if !ok {
 		return err("Math.pow", ErrBadArgument)
 	}
-	y, ok := zngnative.CoerceNativeToFloat64(args[1])
+	y, ok := CoerceToFloat(args.vals[1])
 	if !ok {
 		return err("Math.pow", ErrBadArgument)
 	}
 	r := math.Pow(x, y)
 	if math.IsNaN(r) {
 		return err("Math.pow", ErrBadArgument)
 	}
-	return zngnative.Value{zng.TypeFloat64, r}, nil
+	return zng.Value{zng.TypeFloat64, args.Float64(r)}, nil
 }
 
-func mathSqrt(args []zngnative.Value) (zngnative.Value, error) {
-	var x float64
-	switch args[0].Type.ID() {
-	case zng.IdFloat64:
-		x = args[0].Value.(float64)
-	case zng.IdInt16, zng.IdInt32, zng.IdInt64:
-		x = float64(args[0].Value.(int64))
-	case zng.IdByte, zng.IdUint16, zng.IdUint32, zng.IdUint64:
-		x = float64(args[0].Value.(uint64))
-	default:
+func mathSqrt(args *Args) (zng.Value, error) {
+	x, ok := CoerceToFloat(args.vals[0])
+	if !ok {
 		return err("Math.sqrt", ErrBadArgument)
 	}
-
-	r := math.Sqrt(x)
-	if math.IsNaN(r) {
+	x = math.Sqrt(x)
+	if math.IsNaN(x) {
 		// For now we can't represent non-numeric values in a float64,
 		// we will revisit this but it has implications for file
 		// formats, zql, etc.
 		return err("Math.sqrt", ErrBadArgument)
 	}
-
-	return zngnative.Value{zng.TypeFloat64, r}, nil
+	return zng.Value{zng.TypeFloat64, args.Float64(x)}, nil
 }
 
-func stringByteLen(args []zngnative.Value) (zngnative.Value, error) {
-	switch args[0].Type.ID() {
-	case zng.IdString, zng.IdBstring:
-		v := len(args[0].Value.(string))
-		return zngnative.Value{zng.TypeInt64, int64(v)}, nil
-	default:
+// XXX we should just have a len function that applies to different types
+// and a way to get unicode char len, charlen()?
+func stringByteLen(args *Args) (zng.Value, error) {
+	zv := args.vals[0]
+	if !zng.IsStringy(zv.Type.ID()) {
 		return err("Strings.byteLen", ErrBadArgument)
 	}
-
+	v := len(string(zv.Bytes))
+	return zng.Value{zng.TypeInt64, args.Int(int64(v))}, nil
 }
 
-func stringFormatFloat(args []zngnative.Value) (zngnative.Value, error) {
-	if args[0].Type.ID() != zng.IdFloat64 {
+func stringFormatFloat(args *Args) (zng.Value, error) {
+	zv := args.vals[0]
+	if zv.Type.ID() != zng.IdFloat64 {
 		return err("string.floatToString", ErrBadArgument)
 	}
-	s := strconv.FormatFloat(args[0].Value.(float64), 'g', -1, 64)
-	return zngnative.Value{zng.TypeString, s}, nil
+	f, _ := zng.DecodeFloat64(zv.Bytes)
+	s := strconv.FormatFloat(f, 'g', -1, 64)
+	return zng.Value{zng.TypeString, zng.EncodeString(s)}, nil
```

Comment:
Should scratch have a `String(str string) zcode.Bytes` function?

Reply:
Excellent point.  I think this might be non-trivial so I created #1412 to do the work in a subsequent PR.